File: worker/test/run-evidence-conformance.test.ts
import { readFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { validateRunEvidence } from "../src/run-receipt.ts";

const fixturesPath = join(
  dirname(fileURLToPath(import.meta.url)),
  "../../internal/cli/testdata/run-evidence-conformance/fixtures.json",
);

interface ConformanceFixture {
  name: string;
  evidence: unknown;
  expected: "accept" | "reject";
  error_contains?: string;
}

interface ConformanceFile {
  description: string;
  fixtures: ConformanceFixture[];
}

const file = JSON.parse(readFileSync(fixturesPath, "utf-8")) as ConformanceFile;

// Minimal binding for validation. The conformance corpus tests structural
// validation, not run binding, so we use permissive binding values.
const binding = {
  runID: "",
  leaseID: "",
  provider: "test-provider",
  exitCode: 0,
  receipt: undefined as
    | {
        schema_version: number;
        evidence_sha256?: string;
      }
    | undefined,
};

describe("run-evidence conformance corpus", () => {
  it("should have fixtures", () => {
    expect(file.fixtures.length).toBeGreaterThan(0);
  });

  for (const fx of file.fixtures) {
    it(`fixture: ${fx.name}`, async () => {
      // Adjust binding to match the evidence's exit_code and provider so
      // binding checks don't interfere with structural validation tests.
      const ev = fx.evidence as Record<string, unknown>;
      const localBinding = {
        ...binding,
        provider: typeof ev["provider"] === "string" ? ev["provider"] : "test-provider",
        exitCode: typeof ev["exit_code"] === "number" ? ev["exit_code"] : 0,
        runID: typeof ev["run_id"] === "string" ? ev["run_id"] : "",
        leaseID: typeof ev["lease_id"] === "string" ? ev["lease_id"] : "",
      };

      // For accept fixtures, provide a mock v3 receipt that binds the
      // evidence digest. This allows testing structural validation
      // without needing a real Ed25519-signed receipt.
      if (fx.expected === "accept" && typeof ev["digest"] === "string") {
        localBinding.receipt = {
          schema_version: 3,
          evidence_sha256: ev["digest"],
        };
      }

      const err = await validateRunEvidence(fx.evidence, localBinding);

      if (fx.expected === "reject") {
        expect(err, `expected rejection but evidence was accepted`).toBeInstanceOf(Error);
        if (fx.error_contains && err instanceof Error) {
          expect(
            err.message.toLowerCase(),
            `expected error containing "${fx.error_contains}"`,
          ).toContain(fx.error_contains.toLowerCase());
        }
      } else {
        expect(
          err,
          `expected acceptance but got error: ${err instanceof Error ? err.message : String(err)}`,
        ).toBeUndefined();
      }
    });
  }
});

File: internal/cli/run_evidence_conformance_test.go
package cli

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// conformanceFixture is a single case in the shared Go↔TypeScript evidence
// conformance corpus. Both runtimes must accept or reject each fixture
// identically.
type conformanceFixture struct {
	Name          string          `json:"name"`
	Evidence      json.RawMessage `json:"evidence"`
	Expected      string          `json:"expected"` // "accept" or "reject"
	ErrorContains string          `json:"error_contains,omitempty"`
}

type conformanceFile struct {
	Description string               `json:"description"`
	Fixtures    []conformanceFixture `json:"fixtures"`
}

// TestRunEvidenceConformanceCorpus loads the shared fixture file and verifies
// that the Go evidence parser/validator accepts or rejects each fixture as
// expected. The same fixture file is consumed by the TypeScript test
// (worker/test/run-evidence-conformance.test.ts), and both must agree.
//
// This is the protocol-conformance gate: it is more important than merely
// matching hashes. If Go accepts a fixture that TypeScript rejects (or vice
// versa), the wire contract is broken.
func TestRunEvidenceConformanceCorpus(t *testing.T) {
	path := filepath.Join("testdata", "run-evidence-conformance", "fixtures.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read conformance fixtures: %v", err)
	}
	var file conformanceFile
	if err := json.Unmarshal(data, &file); err != nil {
		t.Fatalf("parse conformance fixtures: %v", err)
	}
	if len(file.Fixtures) == 0 {
		t.Fatal("no conformance fixtures found")
	}
	for _, fx := range file.Fixtures {
		t.Run(fx.Name, func(t *testing.T) {
			// Parse the evidence with the strict parser.
			ev, parseErr := ParseRunEvidenceV1(fx.Evidence)
			// If parsing fails, check whether rejection was expected.
			if parseErr != nil {
				if fx.Expected == "reject" {
					if fx.ErrorContains != "" && !strings.Contains(parseErr.Error(), fx.ErrorContains) {
						t.Errorf("expected error containing %q, got: %v", fx.ErrorContains, parseErr)
					}
					return
				}
				t.Fatalf("unexpected parse error: %v", parseErr)
			}
			// Validate the parsed evidence semantically.
			validateErr := ValidateRunEvidenceV1(ev)
			if fx.Expected == "reject" {
				// Either parse or validation should have rejected it.
				var errStr string
				if validateErr != nil {
					errStr = validateErr.Error()
				} else {
					// If neither parse nor validation rejected it, but the
					// fixture expects rejection, the most common reason is
					// a digest mismatch (the zero digest won't match).
					// Check the digest.
					digestErr := VerifyRunEvidenceDigestError(ev)
					if digestErr != nil {
						errStr = digestErr.Error()
					}
				}
				if errStr == "" {
					t.Fatalf("expected rejection but evidence was accepted")
				}
				if fx.ErrorContains != "" && !strings.Contains(errStr, fx.ErrorContains) {
					t.Errorf("expected error containing %q, got: %s", fx.ErrorContains, errStr)
				}
				return
			}
			// Expected accept: validation should pass.
			if validateErr != nil {
				t.Fatalf("unexpected validation error: %v", validateErr)
			}
		})
	}
}
